### PR TITLE
VMware: Add timeout in wait_for_task

### PR DESCRIPTION
**Pull request by Akasurde**
##### SUMMARY

This fix improves the `wait_for_task` with following improvements - 

* Added exponential backoff algorithm
* Added timeout

Signed-off-by: Abhijeet Kasurde <akasurde@redhat.com>

##### ISSUE TYPE
 - Feature Pull Request

##### COMPONENT NAME
lib/ansible/module_utils/vmware.py

##### ANSIBLE VERSION

```
2.6devel
```

### CODE REVIEW

File: lib/ansible/module_utils/vmware.py [hunk 1]
```diff
@@ -27,19 +28,7 @@ class TaskError(Exception):
 
 
 def wait_for_task(task):
```
Comment (webknjaz) [inline lib/ansible/module_utils/vmware.py]: you can do 
```python
wait_for_task = PyVmomi.wait_for_task
```
or vice versa, but I don't see a point of such duplication aliasing.

Why did you move it to class method in the first place?

File: lib/ansible/module_utils/vmware.py [hunk 2]
```diff
@@ -1110,3 +1099,32 @@ def find_datastore_cluster_by_name(self, datastore_cluster_name):
             if dsc.name == datastore_cluster_name:
                 return dsc
         return None
+
+    @staticmethod
+    def wait_for_task(task, max_backoff=64, timeout=3600):
+        """
+        Function to wait for given task using exponential backoff algorithm
```
Comment (webknjaz) [inline lib/ansible/module_utils/vmware.py]: Could you please use PEP 257 to write docstrings?

File: lib/ansible/module_utils/vmware.py [hunk 3]
```diff
@@ -1110,3 +1099,32 @@ def find_datastore_cluster_by_name(self, datastore_cluster_name):
...
+        Args:
+            task: VMware task object
+            max_backoff: Maximum amount of sleep time in seconds
+            timeout: Timeout for the given task in seconds
+
+        Returns: TaskError for failure or True with result for successful task.
+        """
+        previous_failure = 0
```
Comment (webknjaz) [inline lib/ansible/module_utils/vmware.py]: I'd call it a failure counter. I didn't get the meaning of this var while scanning for the first time.

File: lib/ansible/module_utils/vmware.py [hunk 4]
```diff
@@ -1110,3 +1099,32 @@ def find_datastore_cluster_by_name(self, datastore_cluster_name):
             if dsc.name == datastore_cluster_name:
                 return dsc
         return None
+
+    @staticmethod
+    def wait_for_task(task, max_backoff=64, timeout=3600):
```
Comment (webknjaz) [inline lib/ansible/module_utils/vmware.py]: So how does this method logically relate to the class?

File: lib/ansible/module_utils/vmware.py (hunk 1 above)
Comment (Akasurde) [reply lib/ansible/module_utils/vmware.py]: We are moving every function inside `PyVmomi` class, but we are doing this gradually so once all modules migrate to `PyVmomi` class, we will remove this function all together.

File: lib/ansible/module_utils/vmware.py (hunk 2 above)
Comment (Akasurde) [reply lib/ansible/module_utils/vmware.py]: OK.

File: lib/ansible/module_utils/vmware.py (hunk 3 above)
Comment (Akasurde) [reply lib/ansible/module_utils/vmware.py]: OK.

File: lib/ansible/module_utils/vmware.py (hunk 4 above)
Comment (Akasurde) [reply lib/ansible/module_utils/vmware.py]: Like I said earlier this is part of big task where we move everything under PyVmomi class to avoid duplicity.

File: lib/ansible/module_utils/vmware.py (hunk 1 above)
Comment (webknjaz) [reply lib/ansible/module_utils/vmware.py]: I would suggest changing calling this in this module to calling the static method instead. But on the other side, is it really needed? What's wrong with having utility functions separate from class? They are not related to instances, so it's logical to keep them separate.

**Comment by Akasurde** [comment]
@webknjaz Could you please take another look, I updated patch with review requests ?

**Comment by Akasurde** [comment]
cc @pdellaert @dericcrago

File: lib/ansible/module_utils/vmware.py [hunk 5]
```diff
@@ -29,20 +30,41 @@ class TaskError(Exception):
...
-def wait_for_task(task):
+def wait_for_task(task, max_backoff=64, timeout=3600):
+    """Function to wait for given task using exponential back-off algorithm.
+
+    Args:
+        task: VMware task object
+        max_backoff: Maximum amount of sleep time in seconds
+        timeout: Timeout for the given task in seconds
+
+    Returns: TaskError for failure or True with result for successful task.
```
Comment (webknjaz) [inline lib/ansible/module_utils/vmware.py]: You should state `Raises: TaskError` separately, because it's not a return value

File: lib/ansible/module_utils/vmware.py [hunk 6]
```diff
@@ -29,20 +30,41 @@ class TaskError(Exception):
...
+    gen_errors = (
+        vim.fault.CannotAccessVmConfig,
+        vmodl.fault.InvalidArgument,
+    )
 
     while True:
+        if time.time() - start_time >= timeout:
+            raise TaskError("Timeout")
         if task.info.state == vim.TaskInfo.State.success:
             return True, task.info.result
```
Comment (webknjaz) [inline lib/ansible/module_utils/vmware.py:52]: This is a bad API: you indicate failure/success by both raising exceptions and adding metadata via return value.
It's common in C or Go to use return status codes, but in Python it's just fine to raise exceptions for failures and then return ready-to-use values for successful outcome. It's pointless to always return static `True` additionally to useful payload.

I suggest you to simplify this by wiping out this boolean part and returning just useful result, not a tuple.

File: lib/ansible/module_utils/vmware.py [hunk 7]
```diff
@@ -29,20 +30,41 @@ class TaskError(Exception):
     pass
 
 
-def wait_for_task(task):
+def wait_for_task(task, max_backoff=64, timeout=3600):
+    """Function to wait for given task using exponential back-off algorithm.
```
Comment (webknjaz) [inline lib/ansible/module_utils/vmware.py]: Rephrase: `Wait for given task to finish using exponential back-off algorithm.`
Don't use `Function ...`, because it's useless: the reader can see that it's a function just one line above, when writing useful metadata/comments focus on what might be really important to the end-user of this API. This ensures they'll have a pleasant experience when using your module's API.

File: lib/ansible/module_utils/vmware.py (hunk 6 above)
Comment (Akasurde) [reply lib/ansible/module_utils/vmware.py:52]: I totally agree with your comments, but like I said earlier - this is a legacy function and used by lot of modules. I will need to go and change every module which uses it and test it again. Changing might have detrimental effects rather than doing something good. Let me know if your views on this.

File: lib/ansible/module_utils/vmware.py (hunk 7 above)
Comment (Akasurde) [reply lib/ansible/module_utils/vmware.py]: OK.

File: lib/ansible/module_utils/vmware.py (hunk 6 above)
Comment (webknjaz) [reply lib/ansible/module_utils/vmware.py:52]: Okay, if it's used by lots of modules it should go to the other PR.

**Comment by Akasurde** [comment]
@webknjaz Could you please take another look ?

File: lib/ansible/module_utils/vmware.py [hunk 8]
```diff
@@ -29,20 +30,42 @@ class TaskError(Exception):
...
+        timeout: Timeout for the given task in seconds
+
+    Returns: True with result for successful task.
+    Raises: TaskError on failure
+    """
+    failure_counter = 0
+    start_time = time.time()
+
+    # List of generic errors
+    gen_errors = (
```
Comment (webknjaz) [inline lib/ansible/module_utils/vmware.py]: I think this var needs a more specific name. Like `generic_vmw_errors`, if I understand correctly. Because now it's not clear for me what exactly you're generalizing.

File: lib/ansible/module_utils/vmware.py [hunk 9]
```diff
@@ -29,20 +30,42 @@ class TaskError(Exception):
...
+    )
 
     while True:
+        if time.time() - start_time >= timeout:
+            raise TaskError("Timeout")
         if task.info.state == vim.TaskInfo.State.success:
             return True, task.info.result
         if task.info.state == vim.TaskInfo.State.error:
             try:
+                if isinstance(task.info.error, gen_errors):
```
Comment (webknjaz) [inline lib/ansible/module_utils/vmware.py]: Also, it looks like you could stop relying on exception comparison and  do smth like:
```python
from ansible.module_utils import six
"""Do smth like this in module heading."""
...
error_msg = task.info.error
try:
    error_msg = error_msg.msg
except AttributeError:
    """Keep message as is if it doesn't have `msg` attribute."""
    pass
finally:
    """
    six's `raise_from(exception1, exeption2)` does `raise exception1 from exeption2` under Python 3 where it's supported and will do `raise exception1`, where it's not. 
    """
    six.raise_from(TaskError(error_msg), task.info.error)
```

File: lib/ansible/module_utils/vmware.py [hunk 10]
```diff
@@ -29,20 +30,42 @@ class TaskError(Exception):
...
-def wait_for_task(task):
+def wait_for_task(task, max_backoff=64, timeout=3600):
+    """Wait for given task using exponential back-off algorithm.
+
+    Args:
+        task: VMware task object
+        max_backoff: Maximum amount of sleep time in seconds
+        timeout: Timeout for the given task in seconds
+
+    Returns: True with result for successful task.
```
Comment (webknjaz) [inline lib/ansible/module_utils/vmware.py]: It looks like you put trailing period here, while you use different style for other lines. Could you please keep consistent style? :)

**Review by webknjaz — COMMENTED** [review]
Good job polishing the PR! It looks very clean now!

Though, I would appreciate if you could correct those few nitpicks :)

**Comment by Akasurde** [comment]
@webknjaz Can you please take a look ? thanks.

File: lib/ansible/module_utils/vmware.py [hunk 11]
```diff
@@ -23,26 +24,44 @@
...
-def wait_for_task(task):
+def wait_for_task(task, max_backoff=64, timeout=3600):
+    """Wait for given task using exponential back-off algorithm.
+
+    Args:
+        task: VMware task object
+        max_backoff: Maximum amount of sleep time in seconds
+        timeout: Timeout for the given task in seconds
+
+    Returns: True with result for successful task
```
Comment (webknjaz) [inline lib/ansible/module_utils/vmware.py]: I'd also explicitly state that it returns a tuple. With current wording it seems like it returns just `True` in a certain case.

**Review by webknjaz — APPROVED** [review]
I like it

File: lib/ansible/module_utils/vmware.py (hunk 11 above)
Comment (Akasurde) [reply lib/ansible/module_utils/vmware.py]: Done.

**Comment by webknjaz** [comment]
shipit!

**Comment by Akasurde** [comment]
@webknjaz Thanks for the reviews.
